x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvContains.java

```diff
@@ -0,0 +1,799 @@
+    }
+
+    // @Evaluator(extraName = "Boolean") see end of file.
+    static void process(BooleanBlock.Builder builder, int position, BooleanBlock field1, BooleanBlock field2) {
+        appendTo(builder, containsAll(field1, field2, position, BooleanBlock::getBoolean));
+    }
+
+    // @Evaluator(extraName = "Long") see end of file.
+    static void process(BooleanBlock.Builder builder, int position, LongBlock field1, LongBlock field2) {
+        appendTo(builder, containsAll(field1, field2, position, LongBlock::getLong));
+    }
+
+    // @Evaluator(extraName = "Double") see end of file.
+    static void process(BooleanBlock.Builder builder, int position, DoubleBlock field1, DoubleBlock field2) {
+        appendTo(builder, containsAll(field1, field2, position, DoubleBlock::getDouble));
+    }
+
+    // @Evaluator(extraName = "BytesRef") see end of file.
+    static void process(BooleanBlock.Builder builder, int position, BytesRefBlock field1, BytesRefBlock field2) {
+        appendTo(builder, containsAll(field1, field2, position, (block, index) -> {
+            var ref = new BytesRef();
+            // we pass in a reference, but sometimes we only get a return value, see ConstantBytesRefVector.getBytesRef
+            ref = block.getBytesRef(index, ref);
+            // pass empty ref as null
+            if (ref.length == 0) {
+                return null;
+            }
+            return ref;
+        }));
+    }
+
+    static void appendTo(BooleanBlock.Builder builder, Boolean bool) {
+        if (bool == null) {
+            builder.appendNull();
+        } else {
+            builder.beginPositionEntry().appendBoolean(bool).endPositionEntry();
+        }
+    }
+
+    /**
+     * A block is considered a subset if the superset contains values that test equal for all the values in the subset, independent of
+     * order. Duplicates are ignored in the sense that for each duplicate in the subset, we will search/match against the first/any value
+     * in the superset.
+     *
+     * @param superset block to check against
+     * @param subset   block containing values that should be present in the other block.
+     * @return {@code true} if the given blocks are a superset and subset to each other, {@code false} if not.
+     */
+    static <BlockType extends Block, Type> Boolean containsAll(
+        BlockType superset,
+        BlockType subset,
+        final int position,
+        ValueExtractor<BlockType, Type> valueExtractor
+    ) {
+        if (superset == subset) {
+            return true;
+        }
+        if (subset.areAllValuesNull()) {
+            return true;
+        }
+
+        final var valueCount = subset.getValueCount(position);
+        final var startIndex = subset.getFirstValueIndex(position);
+        for (int valueIndex = startIndex; valueIndex < startIndex + valueCount; valueIndex++) {
+            var value = valueExtractor.extractValue(subset, valueIndex);
```

Comment:
Uhm I didn't see this before, but I think boxing could be a problem performance-wise.

We usually have an evaluator/vector/block per type for 2 reasons:
- Avoid itable accesses
- Avoid primitive boxing

Now, how much extra time will this take, I don't know. This `MV_CONTAINS` function has more overhead than other scalar functions, so maybe it's not that important. But I'm not sure about that.

There are 2 optimizations that would be ideal here:
1. Having a method per type
2. Having a specialization for sorted values (See `Block#mvOrdering()`)

The second one can be made later. The first one too, I guess?

It's in preview, so I think it's fine either way. If it's merged as-is, I would create an issue to improve it later

Reply:
Btw, to do this per-type, whether now or later, this could be a case of using StringTemplates ([Example](https://github.com/elastic/elasticsearch/blob/15c461d8426dacd9f1db4d29a9630ed49ddd0fae/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-ArrayVector.java.st)), to avoid repeated code.
These functions could be extracted into their own static classes (Or the same, but for the full evaluators, with the functions directly inside).

Reply:
yes, should create a followup issue for this. Also want to follow up on improving the Implementers to avoid having Evaluators here and have them autogenerated. Leaving this unresolved until after merge so I don't forget to open an issue.